lib/pbench/server/api/__init__.py

```diff
@@ -120,6 +121,12 @@ def register_endpoints(api, app, config):
         endpoint="datasets_values",
         resource_class_args=(config, logger),
     )
+    api.add_resource(
+        DatasetOwner,
+        f"{base_uri}/datasets/change_owner/<string:dataset>",
```

Comment:
I've thought several times about what this should be *called*, but never really decided. Somehow, though, I don't like the two-part "change owner".
1. On the CLI side we're moving away from verb-noun to noun-verb
2. All of the other API paths are, if not "noun/verb" (like `datasets/publish`) then at least "noun/operator", and the endpoint key name (`datasets_owner`) reflects that path. I really don't like "datasets_change_owner" or "datasets_owner_change", and I don't like the inconsistency between this and the URI path.

I suppose one option is to just remove the "change" verb and go with a "noun/operator" of "datasets/owner". I'm half tempted to suggest "adopt" as a verb, since I've used that term casually, but I don't think that's really what we want.

Reply:
I think that's correct `set` method can be implemented via the `PUT` operation. I also liked `datasets/owner` and we just need to allow `put` and `get` operations on it.

Reply:
I'd rather not introduce `PUT` here. This isn't an update or creation of a "resource" value, it's an "action" on a resource, and that's normally done with `POST`. Supporting `PUT` here would require additional changes. Right now we use `PUT` to upload datasets and to alter/create metadata keys: everything else is `POST` "action" and `GET` "query".

I have considered that `datasets/delete` could be a `DELETE` operation, but we don't really have a "resource" URI. In a "resource oriented" API we'd probably do `PUT` or `POST` to create a `/dataset/<resource_id>` (not an `/upload` "action") and then `DELETE /dataset/<resource_id>`. Even so, I'd be uncomfortable using `PUT` to change individual attributes like owner and access: those are actions, not direct resource updates.

And of course then we wouldn't have to worry about verbs, because our "actions" just identify a resource and one or more action(s) (e.g., `POST /dataset/<resource_id>?access=private&owner=webb`)

This is the "V2 API" I've wanted to work towards, but I really want a better sense of our full API scope, which means figuring out what we're going to do with all the holdover Elasticsearch query APIs that correspond to the arbitrary interactions in the old dashboard rather than anything anyone necessarily actually wants to do with our data.

Reply:
In fact, if we want to get fancy here ... one option would be to enhance the `DatasetsPublish` schema to accept `owner` as well as `access` and change the URI to just `/dataset/<dataset_id>`. As Webb pointed out, there's a lot of duplicate code already. I'd hold off for a separate PR to convert `/datasets/delete` to just be a `DELETE` schema on the same URI, maybe; but this would be a good direction and consolidation.

Note that this would kinda point towards Peter's preference (which I don't actually like) for inventory/contents, where they'd be integrated as just `GET /datasets/<dataset_id>` and the target would be defined by the path: `GET /datasets/<dataset_id>` returns the tarball, `GET /datasets/<dataset_id>/metadata.log` returns the `metadata.log` contents, and `GET /datasets/<dataset_id>/sysinfo` returns directory information for `sysinfo`. (We could specify intent with a required query parameter, though, like `/datasets/<dataset_id>/sysinfo?type=directory` which would fail if `sysinfo` is a regular file; or the converse, `?type=file` which might fail for a directory (or simply return directory metadata without returning the directory contents).

Reply:
> `POST /dataset/<resource_id>?access=private&owner=webb`
> 
> This is the "V2 API" I've wanted to work towards

Yes, this is presumably what is in the back of Nikhil's and my minds...except my understanding is that would be a `PUT` if we're modifying them (otherwise, the query parameters would be specifying qualifiers for an upload...).

> one option would be to enhance the `DatasetsPublish` schema to accept owner as well as access and change the URI to just `/dataset/<dataset_id>`. As Webb pointed out, there's a lot of duplicate code already. 

This sounds good to me.  Yeah, we don't need to jump on `DELETE` or the inventory/contents stuff, but adding ownership change in a way that moves us towards V2 would be good

Reply:
In REST terms, a `PUT` creates or replaces a resource, as our existing `PUT` provides an entire dataset tarball. What we're doing in both `publish` and "owner change" could conceivably be considered a `PATCH` operation on the resource, but definitely not a `PUT`. It's an *action* against a resource, which is a `POST` with action name and parameters. (And I don't even want to get into defining and implementing `PATCH` at this point, though that might become a useful model in the future.)

But I do like the idea of merging publish/owner/delete into a single class with a `PUT` schema having `dataset_id` URI parameter and `owner` and `access` query parameters, along with a `DELETE` schema that just takes the URI `resource_id`. That would be cool.

And it would be fine to start by just merging this change with `publish` into a `PUT /datasets/<dataset_id>` URI... we can fold in `DELETE` later. In fact, `/datasets/list` could become a query on the collection, as `GET /datasets`, leaving `GET /datasets/<dataset_id>[path]` for contents and inventory ... although we need to think carefully about exactly how we merge them. (E.g., how much the context is implicit.)

Reply:
> But I do like the idea of merging publish/owner/delete into a single class with a `PUT` schema having `dataset_id` URI parameter and `owner` and `access` query parameters, along with a `DELETE` schema that just takes the URI `resource_id`.

Yeah . This is an excellent approach moving forward.

>And it would be fine to start by just merging this change with publish into a PUT /datasets/<dataset_id> URI... we can fold in DELETE later.

Yes. For this PR, I'll start with combining `publish` and `owner` API  functionality into one with `PUT /datasets/<dataset_id>` URI

Reply:
> Yes. For this PR, I'll start with combining `publish` and `owner` API functionality into one with `PUT /datasets/<dataset_id>` URI

`POST`, not `PUT`, please!